src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -361,6 +366,217 @@ namespace cryptonote
+            m_core.handle_incoming_tx(tx_blob, tvc, true, true);
+            
+            if(tvc.m_verifivation_failed)
+            {
+              LOG_PRINT_CCONTEXT_L1("Block verification failed: transaction verification failed, dropping connection");
+              m_p2p->drop_connection(context);
+              m_core.resume_mine();
+              return 1;
+            }
+            
+            // Should only not be added to pool if verification failed, but
+            // maybe in the future could not be added for other reasons 
+            // according to monero-moo so keep track of these separately ..
+            //
+            // this is probably the wrong thing to do, but for now the idea
+            // would be to pass these back/forth on the missing tx request,
+            // right now they're just being lost at the end of this function call
+            if(!tvc.m_added_to_pool) 
+              tx_hash_to_blob[tx_hash] = tx_blob;
+          }
+        }
+        else
+        {
+          LOG_ERROR_CCONTEXT
+          (
+            "sent wrong tx: failed to parse and validate transaction: \r\n"
+            << epee::string_tools::buff_to_hex_nodelimer(tx_blob) 
+            << "\r\n dropping connection"
+          );
+            
+          m_p2p->drop_connection(context);
+          m_core.resume_mine();
+          return 1;
+        }
+      }
+        
+      BOOST_FOREACH(auto& tx_hash, new_block.tx_hashes)
+      {
+        if(m_core.get_pool_transaction(tx_hash, tx))
+        {
+            have_tx.push_back(tx_to_blob(tx));
+        }
+        else
+        {
+          if(tx_hash_to_blob.find(tx_hash) != tx_hash_to_blob.end()) 
+          {
+              have_tx.push_back(tx_hash_to_blob[tx_hash]);
+          }
+          else
+          {
+            need_tx_hashes.push_back(tx_hash); 
+          }
+        }
+      }
+        
+      if(need_tx_hashes.size() > 0) // drats, we don't have everything..
+      {
+        // request non-mempool txs
+        NOTIFY_REQUEST_FLUFFY_MISSING_TX::request missing_tx_req;
+        missing_tx_req.b = arg.b;
+        missing_tx_req.hop = arg.hop;
+        BOOST_FOREACH(auto& tx_hash, need_tx_hashes)
+        {
+          missing_tx_req.missing_tx_hashes.push_back(tx_hash);
+        }
+
+        m_core.resume_mine();
+        post_notify<NOTIFY_REQUEST_FLUFFY_MISSING_TX>(missing_tx_req, context);
+      }
+      else // whoo-hoo we've got em all ..
+      {
+        block_complete_entry b;
+        b.block = arg.b.block;
+        b.txs = have_tx;
+
+        std::list<block_complete_entry> blocks;
+        blocks.push_back(b);
+        m_core.prepare_handle_incoming_blocks(blocks);
+
+        // Do I need to do this? Every transaction is either already in the pool
+        // and thus verified(?), or I just verified it above if it was included
+        // in the sent txs. So is this a double check / wasteful?
```

Comment:
I guess you could always ask before each send, although there is overhead in that and I personally think the swap capabilities on handshake model is better.